README.md
# MyApnea

[![Build Status](https://travis-ci.com/myapnea/myapnea.org.svg?branch=master)](https://travis-ci.com/myapnea/myapnea.org)
[![Code Climate](https://codeclimate.com/github/myapnea/myapnea.org/badges/gpa.svg)](https://codeclimate.com/github/myapnea/myapnea.org)

The web framework behind https://myapnea.org, built Rails 6.1+ and Ruby 3.1+.

## Before Starting Installation

Make sure you have reviewed and installed any
[prerequisites](https://github.com/myapnea/myapnea.org/blob/master/PREREQUISITES.md).

## Installation

```
gem install bundler
```

This README assumes the following installation directory:
`/var/www/myapnea.org`

```
cd /var/www

git clone https://github.com/myapnea/myapnea.org.git

cd myapnea.org

bundle install
```

Install default configuration files for database connection, email server
connection, server url, and application name.

```
ruby lib/initial_setup.rb

rails db:migrate RAILS_ENV=production

rails assets:precompile RAILS_ENV=production
```

Edit the `/config/application.yml` file it generates, and enter any keys you've
gotten from the third party developers. For any service you aren't using, you
can leave the lines out.

Run Rails Server (or use Apache or nginx)

```
rails s
```

Open a browser and go to: [http://localhost:3000](http://localhost:3000)

All done!

## Refreshing Sitemap

Edit Cron Jobs `sudo crontab -e` to run the task `lib/tasks/sitemap.rake`

```
SHELL=/bin/bash
0 1 * * * source /etc/profile.d/rvm.sh && cd /var/www/myapnea.org && rvm 3.1.2 && rails sitemap:refresh RAILS_ENV=production
```

## License

MyApnea is released under the [MIT License](http://www.opensource.org/licenses/MIT).


CHANGELOG.md
## 30.0.1 (July 17, 2022)

### Enhancements
- **General Changes**
  - Removed links to research pages

### Bug Fix
- Fixed issue with maintaining text captcha responses across multiple servers

## 30.0.0 (June 18, 2022)

### Enhancements
- **General Changes**
  - Added a text captcha to the registration page
- **Resource Changes**
  - Admins can now attach PDFs to resources
- **Gem Changes**
  - Update to rails 6.1.6

## 29.2.0 (May 7, 2022)

### Bug Fix
- Fixed landing page text overflow issue on small screens

## 29.1.0 (April 23, 2022)

### Enhancements
- **Security Changes**
  - Fixed a Cross Site Scripting (XSS) issue
- **Gem Changes**
  - Update to Ruby 3.1.2

## 29.0.0 (April 19, 2022)

### Enhancements
- **General Changes**
  - Updated landing page based on new design
  - The research page is now fixed with for the public facing view
  - Improved consistency of styling across various public pages
  - Added Fathom Analytics to website and removed Google Analytics
  - Updated internal dashboard pages to match new theme
- **Resource Page Changes**
  - Resources now have a slug that can act as a short link to the resource
  - Resources can now be configured to open in a new tab
- **Gem Changes**
  - Update to ruby 3.0.2

## 28.0.0 (November 7, 2021)

### Enhancements
- **Gem Changes**
  - Update to rails 6.1.4.1

### Bug Fix
- Fixed a bug that prevented posts from posting correctly
- Fixed an issue where mobile safari would play the globe video when clicking
  on the MyApnea menu item
- Fixed a bug that would generate excessive notifications for users who chose to
  be auto-subscribed to new forum posts

## 27.0.0 (October 9, 2021)

### Enhancements
- **Forum Changes**
  - Email notifications for replies to forum topics have been re-added
    - Notifications are only sent to users who have subscribed to the forum
      topic and who have emails enabled
  - Old topics are now auto-locked after 100 years (and not 2 months)

### Bug Fix
- Fixed a bug that prevented a nested reply from being linked
- Fixed a bug that caused remote forms from correctly rendering the response
- Fixed video not playing on landing page

## 26.0.0 (April 18, 2021)

### Enhancements
- **General Changes**
  - Privacy policy now redirects to https://www.massgeneralbrigham.org/notices/web-privacy-policy
  - Added Google Analytics
- **Resource Page Changes**
  - Images on resources now also link to the resource URL

### Bug Fix
- Fixed a bug that prevented new articles from reusing a URL slug from a
  previously deleted article

## 25.0.0 (April 11, 2021)

### Enhancements
- **Admin Changes**
  - Add a Slice Subject ID export task
  - Topics marked as spam are no longer listed on the index for admins
- **General Changes**
  - A new resources page was added that allows admins to create and list
    resources for general visitors of the website
  - Added note to the MyApnea contact page that MyApnea is not affiliated with
    MyAir or Resmed
- **Forum Changes**
  - Existing replies to locked topics now be edited or deleted after the topic
    has been locked
- **Gem Changes**
  - Update to ruby 3.0.1
  - Update to rails 6.1.3.1
  - Update to pg 1.2.3
  - Update to bootstrap 4.4.1
  - Update to carrierwave 2.2.1
  - Update to devise 4.7.3
  - Update to figaro 1.2.0
  - Update to font-awesome-sass 5.12.0
  - Update to haml 5.2.1
  - Update to kaminari 1.2.1
  - Update to pg_search 2.3.2
  - Update to redcarpet 3.5.1
  - Update to rubyzip 2.3.0
  - Update to coffee-rails 5.0
  - Update to jquery-rails 4.3.5

### Bug Fixes
- Dashboard forum activity no longer displays replies to deleted topics

## 24.0.1 (May 28, 2019)

### Enhancements
- **Article Changes**
  - Articles in draft mode now display category for editors
- **Gem Changes**
  - Update to haml 5.1.1
  - Update to jbuilder 2.9
  - Update to pg_search 2.2.0
  - Update to rubyzip 1.2.3

### Bug Fix
- Fix an issue displaying topics without last reply at timestamp

## 24.0.0 (May 10, 2019)

### Enhancements
- **Admin Changes**
  - Include "Email confirmed" in user export
- **Gem Changes**
  - Update to ruby 2.6.3
  - Update to devise 4.6.2

## 23.0.0 (April 10, 2019)

### Enhancements
- **Gem Changes**
  - Update to ruby 2.6.2
  - Update to rails 6.0.0.beta3
  - Update to font-awesome-sass 5.8.1

## 22.0.0 (February 26, 2019)

### Enhancements
- **General Changes**
  - Landing page survey count now includes newly completed surveys from Slice
    along with original survey count
  - Tabs now display better on smaller screens
- **Project Changes**
  - A primary project can now be specified that determines the consent displayed
    on the root consent URL
- **Survey Changes**
  - Dashboard now displays next survey question
- **Gem Changes**
  - Update to ruby 2.6.1
  - Update to rails 6.0.0.beta1
  - Update to pg 1.1.4
  - Update to bootstrap 4.3.1
  - Update to carrierwave 1.3.1
  - Update to devise 4.6.1
  - Update to hashids 1.0.5
  - Update to mini_magick 4.9.3
  - Update to pg_search 2.1.4
  - Update to sitemap_generator 6.0.2

## 21.1.0 (January 2, 2019)

### Enhancements
- **General Changes**
  - Contact page support email information moved above support email
- **Gem Changes**
  - Update to ruby 2.6.0
  - Update to rails 5.2.2
  - Update to bootstrap 4.2.1
  - Update to font-awesome-sass 5.6.1

### Bug Fix
- Fixed styling of sunset decoration on admin pages
- Fixed issue caused by replying to restored forum topics that had no slug

## 21.0.0 (November 29, 2018)

### Enhancements
- **Forum Changes**
  - Improved spam filters
- **Survey Changes**
  - Improved process of leaving and rejoining a research study
  - Overview reports and project consents are now generated in temporary
    folders that are deleted after generating the PDFs
- **Gem Changes**
  - Update to rails 5.2.1.1
  - Update to font-awesome-sass 5.5.0

### Bug Fix
- Fixed missing gradient on landing page "Community Power" section

## 20.0.0 (November 5, 2018)

### Enhancements

gems.rb
# frozen_string_literal: true

# rubocop:disable Layout/ExtraSpacing
source "https://rubygems.org"

gem "rails",                      "6.1.6"

# PostgreSQL as the Active Record database.
gem "pg",                         "1.2.3"

# Gems used by project.
gem "acts_as_textcaptcha",        "~> 4.6.0"
gem "autoprefixer-rails"
gem "bootstrap",                  "~> 4.4.1"
gem "carrierwave",                "~> 2.2.1"
gem "devise",                     "~> 4.7.3"
gem "figaro",                     "~> 1.2.0"
gem "font-awesome-sass",          "~> 5.12.0"
gem "haml",                       "~> 5.2.1"
gem "hashids",                    "~> 1.0.5"
gem "jquery-ui-rails",            "~> 6.0.1"
gem "kaminari",                   "~> 1.2.1"
gem "mini_magick",                "~> 4.9.3"
gem "net-imap",                   require: false
gem "net-pop",                    require: false
gem "net-smtp",                   require: false
gem "pg_search",                  "~> 2.3.2"
gem "redcarpet",                  "~> 3.5.1"
gem "rubyzip",                    "~> 2.3.0"
gem "sitemap_generator",          "~> 6.0.2"
gem "activesupport_cache_database", "~> 0.2.1"

# Rails defaults.
gem "coffee-rails",               "~> 5.0"
gem "sass-rails",                 ">= 6"
gem "uglifier",                   ">= 1.3.0"

gem "jbuilder",                   "~> 2.9"
gem "jquery-rails",               "~> 4.3.5"
gem "turbolinks",                 "~> 5"

group :development do
  gem "listen",                   "~> 3.3"
  gem "spring"
  gem "spring-watcher-listen",    "~> 2.0.0"
  gem "web-console",              ">= 4.1.0"
end

# Testing
group :test do
  gem "artifice"
  gem "capybara",                 ">= 3.26"
  gem "minitest"
  gem "puma"
  gem "selenium-webdriver"
  gem "simplecov",                "~> 0.16.1", require: false
end
# rubocop:enable Layout/ExtraSpacing
